### ISSUE TITLE
Fine grained fastmath/NaN+Inf handling.

### ISSUE THREAD

**Issue by stuartarchibald** [numba/numba#2923]
## Feature request

The `fastmath` option to `njit` would benefit from:

1. Being tunable.
2. Having some default combinations for `fast, but still handles floating point behaviour XYZ like IEEE754`.
3. Having warnings printed when using functions that are known to violate some combination (also requested here https://github.com/numba/numba/issues/2919#issuecomment-385322414).

The space of configuration is, IR operations `fadd, fsub, fmul, fdiv, frem, fcmp` and `call`, and flags: `nnan, ninf, nsz, arcp, contract, afn, reassoc` and `fast` which is `apply all flags`.

As a first improvement, the `fastmath` kwarg to `jit` could handle configuration as follows:
* `fastmath = True/False`, the current behaviour, applies all or none of the fast math flags respectively.
* `fastmath = {'nnan' = True, 'ninf' = True, ...}`, a dictionary of flags is handled, (defaults to all True as per `fast`), this is passed to the fastmath rewrite pass to selectively apply the fast flags.
* `fastmath = "fastwithnan"`, this option applies the fast math flags as per the dictionary above, but does not apply `nnan`, this should allow NaN handling to work and the use of `NaN` as a place holder/signal.
* `fastmath = "fastwithinf"`, this option applies the fast math flags as per the dictionary above, but does not apply `ninf`, this should allow Inf handling to work and the use of `Inf` as a place holder/signal.

This should allow the user to switch on fastmath, switch on fastmath but also use NaN/Inf, and also fine tune which flags are applied. I think the first iteration should not involve permitting the tuning of the operations to which each flag set is applied, this is quite complicated and a somewhat niche use case.

Following this change the functions `np.isnan` and `np.isinf` should be augmented to issue a warning if they are used with flags that will render their operation as a NOP/perhaps surprising! There's also a number of functions like `np.nanmean` which would need looking at with view of applying similar depending on whether or not the implementation goes via path already containing a warning.

**Comment by leohaim** [comment]
Thanks for raising this. 

'fastmath'='fastwithnan' or the flags dictionary would work in my use cases. 

numpy.nan?? functions should not be allowed at all inside a function containing "fastmath=True" in the decorator, since this combination breaks semantics. An error instead of a warning might therefore be more appropriate in this case.

**Comment by seibert** [comment]
Yeah, I would agree that we want a way for functions to detect and error out on certain compile flag combinations.  Would a generalization of "generated_jit" that also passed in compiler flags make sense?  This would also allow functions to pick different implementations if things like fastmath (or even parallel accelerator) are active.

**Comment by seibert** [comment]
Siu counter-proposed having the compile flags be accessible as compile-time constants inside the function, but we'd also need a "abort_compilation" function.

**Comment by stuartarchibald** [comment]
For reference, in `C` with GCC, using the inlinable `__buildin_isnan` in a condition like (call with arg `-1`):
```C
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char *argv[])
{
    double x, y;
    if (argc != 2) return -1;
    y = (double)atoi(argv[1]);
    x = sqrt(y);
    printf("%f\n", x);
    if (__builtin_isnan(x))
    {
        printf("x is glibc nan\n");
    }
    return 0;
}
```
Without `-ffast-math` enabled a `ucomisd` is issued with expected branches, with `-ffast-math` enabled the condition and branches are entirely optimised out. If the `glibc` function `isnan` is used instead, this is preserved as a function call irrespective of flags applied.

I'm unsure what behaviour is most desirable.
* Aborting compilation if a flag combination is mismatched with some part of a piece of code is not something I'd expect from a compiler, in spite of it producing "unexpected" results. Issuing a warning as noted may be a way to help with this.
* Perhaps GCC have it right in that in most cases users will use a `isnan` which will "do the right thing" irrespective of flags, but if wanted there is an inlinable version which can be made to respect the flags, even if the result may be considered surprising. (Is this similar to what you are suggesting with the `generated_jit` generalisation above @seibert ?).

In terms of the Numba code base, perhaps we ought to check that the `np.nanFOO` like functions are all written in terms of `np.isnan` and look for cases where there is an `isinf` or `isnan` like behaviour so that the impact of altering compilation of these functions is known. I suspect that some C99 function re-implementations are also going to be affected as these are often written using builder IR and inf/nan traps will likely be optimised out under `fastmath` due to inlining etc.

**Comment by stuartarchibald** [comment]
Closing, implemented in #3847
